pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/RoundRobinScheduler.java

```diff
@@ -39,168 +47,208 @@
-      _available.add(new AvailableEntry(operatorChain, releaseTs));
+    } finally {
+      _lock.unlock();
     }
     trace("registered " + operatorChain);
   }
 
   @Override
-  public void onDataAvailable(MailboxIdentifier mailbox) {
-    // it may be possible to receive this callback when there's no corresponding
-    // operator chain registered to the mailbox - this can happen when either
-    // (1) we get the callback before the first register is called or (2) we get
-    // the callback while the operator chain is executing. to account for this,
-    // we just store it in a set of seen mail and only check for it when hasNext
-    // is called.
-    //
-    // note that scenario (2) may cause a false-positive schedule where an operator
-    // chain gets scheduled for mail that it had already processed, in which case
-    // the operator chain will simply do nothing and get put back onto the queue.
-    // scenario (2) may additionally cause a memory leak - if onDataAvailable is
-    // called with an EOS block _while_ the operator chain is executing, the chain
-    // will consume the EOS block and computeReady() will never remove the mailbox
-    // from the _seenMail set.
-    //
-    // TODO: fix the memory leak by adding a close(opChain) callback
-    _seenMail.add(mailbox);
-    trace("got mail for " + mailbox);
+  public void deregister(OpChain operatorChain) {
+    Preconditions.checkState(_aliveChains.containsKey(operatorChain.getId()),
+        "Tried to de-register an un-registered op-chain");
+    _lock.lock();
+    try {
+      OpChainId chainId = operatorChain.getId();
+      _aliveChains.remove(chainId);
+      // it could be that the onDataAvailable callback was called when the OpChain was executing, in which case there
+      // could be a dangling entry in _seenMail.
+      _seenMail.remove(chainId);
+    } finally {
+      _lock.unlock();
+    }
   }
 
   @Override
-  public boolean hasNext() {
-    if (!_ready.isEmpty()) {
-      return true;
+  public void yield(OpChain operatorChain) {
+    long releaseTs = _ticker.get() + _releaseTimeoutMs;
+    _lock.lock();
+    try {
+      // It could be that this OpChain received data before it could be yielded completely. In that case, mark it ready
+      // to get it scheduled asap.
+      if (_seenMail.contains(operatorChain.getId())) {
+        _seenMail.remove(operatorChain.getId());
+        _ready.add(operatorChain);
+        return;
+      }
+      _available.put(operatorChain.getId(), releaseTs);
+    } finally {
+      _lock.unlock();
     }
-    computeReady();
-    return !_ready.isEmpty();
   }
 
   @Override
-  public OpChain next() {
-    OpChain op = _ready.poll();
-    trace("Polled " + op);
-    return op;
+  public void onDataAvailable(MailboxIdentifier mailbox) {
+    // TODO: Should we add an API in MailboxIdentifier to get the requestId?
+    OpChainId opChainId = new OpChainId(Long.parseLong(mailbox.getJobId().split("_")[0]),
```

Comment:
add a method `getRequestId` in MailboxIdentifier?

Reply:
Yeah so I also wanted us to revisit `MailboxIdentifier::getJobId`. Right now it is `requestId_stageid` but in case of multiple brokers there can be duplicate requests with the same id in a server. Filed: https://github.com/apache/pinot/issues/10305